### PR TITLE
Added benchmarks for optimal synthesis paper.

### PR DESCRIPTION
**Pull request by pehamTom**
## Description

This PR adds benchmarks for the optimal synthesis paper to the repo. If this should be handled another way, please tell me so and I will adapt accordingly.

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [x] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: benchmarks/optimal_synthesis_of_clifford_circuits_with_sat_solvers/grover_3_0.qasm [hunk 1]
```diff
@@ -0,0 +1,11 @@
+OPENQASM 2.0;
+include "qelib1.inc";
+gate ccz q0,q1,q2 { p(pi/4) q1; p(pi/4) q2; cx q1,q2; u(0,0,-pi/4) q2; cx q1,q2; u(0,0,0) q2; cx q1,q0; p(-pi/4) q0; p(pi/4) q2; cx q0,q2; u(0,0,pi/4) q2; cx q0,q2; u(0,-pi/2,0) q2; cx q1,q0; p(pi/4) q0; p(pi/4) q2; cx q0,q2; u(0,0,-pi/4) q2; cx q0,q2; u(0,0,0) q2; }
+gate cz_o0 q0,q1 { x q0; cz q0,q1; x q0; }
+gate gate_Q q0,q1,q2 { cz_o0 q2,q0; x q1; z q1; x q1; ccz q0,q1,q2; cz_o0 q2,q1; h q2; h q1; h q0; x q0; x q1; x q2; h q2; ccx q0,q1,q2; h q2; x q0; x q1; x q2; h q0; h q1; h q2; }
+gate gate_Q q0,q1,q2 { gate_Q q0,q1,q2; }
```
[the remark this answers is not in the document]
Comment (burgholzer) [reply benchmarks/optimal_synthesis_of_clifford_circuits_with_sat_solvers/grover_3_0.qasm:6]: Could you maybe regenerate the files with the newest qiskit terra version (>=0.24.0)? The QASM import and export got significantly revamped recently and should be way faster and better now.

[the remark this answers is not in the document]
Comment (pehamTom) [reply benchmarks/optimal_synthesis_of_clifford_circuits_with_sat_solvers/grover_3_0.qasm]: I have to admit I'm quite impressed with how broken the qasm export truly was. There were also some gate definitions that occured out of order. Fixed it now.

**Review by burgholzer — APPROVED** [review]
Some of the files still seem suboptimal, but that's not on you. Quite some files contain duplicate gate definitions with some cryptic identifier appended. It would be nice to get a minimal reproducer and file a bug report in Qiskit.
Would it be worth it to also add the Python file for generating the benchmarks to the directory so that we can regenerate at any point in time?

File: benchmarks/optimal_synthesis_of_clifford_circuits_with_sat_solvers/grover_3_0.qasm (hunk 1 above)
[the remark this answers is not in the document]
Comment (burgholzer) [reply benchmarks/optimal_synthesis_of_clifford_circuits_with_sat_solvers/grover_3_0.qasm]: See the general comment. It's still not perfect, but looks better.

**Pull request merged**